### ISSUE TITLE
🐛 Bug Report: ConfigLocationProvider executed in read only mode 

### ISSUE THREAD

**Issue by nodify-at** [backstage/backstage#15077]
### 📜 Description

We have created two services in our backstage environment, 1 service is a standalone service to scan to catalog entities and the other one is responsible for backend services.

In our backend we are creating a router using catalog-backend to be able to call backend services of catalog endpoint and we don't trigger the processingEngine (using start function).

https://github.com/backstage/backstage/blob/db71add9ffac4a31dd66ee464fd0466f2bb7ba89/plugins/catalog-backend/src/service/CatalogBuilder.ts#L537

CatalogBuilder triggers connection immediately while building the engine. It causes an inconsistency in our environment. Because we are configuring the locations in our catalog scanner service and keeping it empty in our backend service.

If we start our catalog scanner first of all and after than starting the backend service, location based entities are removed.

This issue happens even if we set readonly: true in our backend. 

### 👍 Expected behavior

1. I expect in readonly mode: nothing should be changed, easiest way would be to trigger  the line below if readonly is false or not set. https://github.com/backstage/backstage/blob/db71add9ffac4a31dd66ee464fd0466f2bb7ba89/plugins/catalog-backend/src/service/CatalogBuilder.ts#L537 
3. or at least: If we don't start the processing engine, the default providers should not execute any kind of catalog processing.
4. ConfigLocationEntityProvider can implement a TaskRunner additional like other providers to sync locations in a scheduled time (if not implemented yet)

I can prepare a PR, if it is OK for you :)

### 👎 Actual Behavior with Screenshots

It actually changes/removes location based catalog files even in readonly mode and processing engine is not executed. It causes to inconsistency in a scaled env.

### 👟 Reproduction steps

1. Prepare two backend applications
3. One with locations, one without locations and readonly: true
4. Start first the application with locations
5. Wait a while and start the second application, it will remove existing entities

### 📃 Provide the context for the Bug.

Related links:

https://github.com/backstage/backstage/blob/db71add9ffac4a31dd66ee464fd0466f2bb7ba89/plugins/catalog-backend/src/modules/core/ConfigLocationEntityProvider.ts#L34

https://github.com/backstage/backstage/blob/db71add9ffac4a31dd66ee464fd0466f2bb7ba89/plugins/catalog-backend/src/service/CatalogBuilder.ts#L537

### 🖥️ Your Environment

OS:   Darwin 22.1.0 - darwin/x64
node: v14.19.0
yarn: 3.2.3
cli:  0.21.1 (installed)
backstage:  1.8.2

Dependencies:
  @backstage/app-defaults                               1.0.8
  @backstage/backend-app-api                            0.2.3
  @backstage/backend-common                             0.16.0
  @backstage/backend-plugin-api                         0.1.4
  @backstage/backend-tasks                              0.3.7
  @backstage/backend-test-utils                         0.1.30
  @backstage/catalog-client                             1.1.2
  @backstage/catalog-model                              1.1.3
  @backstage/cli-common                                 0.1.10
  @backstage/cli                                        0.21.1
  @backstage/config-loader                              1.1.6
  @backstage/config                                     1.0.4
  @backstage/core-app-api                               1.2.0
  @backstage/core-components                            0.11.2, 0.12.0
  @backstage/core-plugin-api                            1.1.0
  @backstage/dev-utils                                  1.0.8
  @backstage/errors                                     1.1.3
  @backstage/integration-react                          1.1.6
  @backstage/integration                                1.4.0
  @backstage/plugin-api-docs                            0.8.11
  @backstage/plugin-auth-backend                        0.17.1
  @backstage/plugin-auth-node                           0.2.7
  @backstage/plugin-badges-backend                      0.1.32
  @backstage/plugin-badges                              0.2.35
  @backstage/plugin-bazaar-backend                      0.2.1
  @backstage/plugin-bazaar                              0.2.0
  @backstage/plugin-catalog-backend-module-github       0.2.1
  @backstage/plugin-catalog-backend                     1.5.1
  @backstage/plugin-catalog-common                      1.0.8
  @backstage/plugin-catalog-graph                       0.2.23
  @backstage/plugin-catalog-import                      0.9.1
  @backstage/plugin-catalog-node                        1.2.1
  @backstage/plugin-catalog-react                       1.2.1
  @backstage/plugin-catalog                             1.6.1
  @backstage/plugin-code-coverage                       0.2.4
  @backstage/plugin-explore-react                       0.0.23
  @backstage/plugin-explore                             0.3.42
  @backstage/plugin-github-actions                      0.5.11
  @backstage/plugin-github-deployments                  0.1.42
  @backstage/plugin-github-pull-requests-board          0.1.5
  @backstage/plugin-graphiql                            0.2.43
  @backstage/plugin-home                                0.4.27
  @backstage/plugin-jenkins-backend                     0.1.28
  @backstage/plugin-jenkins-common                      0.1.10
  @backstage/plugin-jenkins                             0.7.10
  @backstage/plugin-kafka                               0.3.11
  @backstage/plugin-lighthouse                          0.3.11
  @backstage/plugin-org                                 0.6.1
  @backstage/plugin-permission-backend                  0.5.13
  @backstage/plugin-permission-common                   0.6.4, 0.7.1
  @backstage/plugin-permission-node                     0.7.1
  @backstage/plugin-permission-react                    0.4.7
  @backstage/plugin-playlist-backend                    0.2.1
  @backstage/plugin-playlist-common                     0.1.2
  @backstage/plugin-playlist                            0.1.2
  @backstage/plugin-proxy-backend                       0.2.32
  @backstage/plugin-scaffolder-backend                  1.8.0
  @backstage/plugin-scaffolder-common                   1.2.2
  @backstage/plugin-scaffolder                          1.8.0
  @backstage/plugin-search-backend-module-elasticsearch 1.0.4
  @backstage/plugin-search-backend-node                 1.0.4
  @backstage/plugin-search-backend                      1.1.1
  @backstage/plugin-search-common                       0.3.4, 1.1.1
  @backstage/plugin-search-react                        1.2.1
  @backstage/plugin-search                              1.0.4
  @backstage/plugin-shortcuts                           0.3.3
  @backstage/plugin-sonarqube-backend                   0.1.3
  @backstage/plugin-sonarqube                           0.5.0
  @backstage/plugin-stack-overflow                      0.1.7
  @backstage/plugin-tech-radar                          0.5.18
  @backstage/plugin-techdocs-backend                    1.4.1
  @backstage/plugin-techdocs-module-addons-contrib      1.0.6
  @backstage/plugin-techdocs-node                       1.4.2
  @backstage/plugin-techdocs-react                      1.0.6
  @backstage/plugin-techdocs                            1.4.0
  @backstage/plugin-user-settings-backend               0.1.2
  @backstage/plugin-user-settings                       0.5.1
  @backstage/release-manifests                          0.0.7
  @backstage/search-common                              0.3.4
  @backstage/test-utils                                 1.2.2
  @backstage/theme                                      0.2.16
  @backstage/types                                      1.0.1
  @backstage/version-bridge                             1.0.2

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

**Comment by Rugvip** [comment]
Oh that's a bit tricky. On one hand it could make sense to not connect any providers until `engine.start()` is called, but that would prevent you from running dedicated provider ingestion nodes. Don't know how big of a loss that is, but it makes me want to look for other options.

The `readonly` config flag is a bit different than this tbh, it toggles whether users are allowed to register location entities or not, any configured locations will still be used and we'll want to keep it that way.

One workaround for your case is of course to have the same location configuration in your backend service as for the location scanner.

I'm wondering if the best solution here might be to stop defaulting `catalog.locations` in config to `[]`. If the locations are left empty, the behavior could instead be to do nothing? You should be able to override that in additional config files too using `catalog.locations: null`

**Comment by Rugvip** [comment]
> @Rugvip thanks for your answer :) I think we should rename this read only to something like disable location analyser?

It's been confused in multiple different ways before, so I think a rename could make sense x)

> Or a general solution would be to provide an option to replace it deactivate the default providers so we can decide it where/which service enables them explicitly ?

Yep, this is on the radar. We might not try to tackle that before #13081 has gotten further though. i.e. have a solution that's only for the new system

> Another issue (or a feature) would be to improve this provider to run in a scheduled time to sync the locations?actually it runs only once at start up right ?

It subscribes to config updates, which in the default setup means it's use file watch to react to changes in the configuration.

**Comment by nodify-at** [comment]
@Rugvip Thanks for your response :) Until the new backend system publicly available, can we do the following? : 

Add new setter function in builder to replace config location provider, so we can either implement our own solution or remove it so it won't be triggered?

**Comment by Rugvip** [comment]
@nodify-at would working around it by adding your own empty provider with the same name as the location provider be an alright temporary solution? Not feeling we necessarily need to add a setter for that. The worry with adding a setter would be if people use it while still expecting location configs to work e.g.

**Comment by Rugvip** [comment]
I think the `catalog.locations: null` would be a sane approach to this as well, would argue that it's fine to ship that without calling it a breaking change too.
